### ISSUE TITLE
Throttle number of Lambdas deployed at once

### ISSUE THREAD

**Issue by ryanblock** [arc-repos/arc-workflows#110]
This is a followup / partial duplicate of #76 – lots of clever solutions outlined in that that would definitely improve reliability of large arc deploys.

I'm opening this issue to single out what I think/hope is probably the easiest, lowest hanging fruit to move this forward – that being to throttle the number of Lambdas being deployed at once.

Deploying in batches of 10 (say, with a 1000ms pause between sets) makes sense and would likely not have a material impact on most deploy times. But to alleviate potential concerns, we could activate throttling via an `@aws` section option, i.e.:
```
@aws
profile work
region us-west-1
throttle lambda
```
(To further that idea, `throttle` could also be used for something like API Gateway, which we've also seen be subject to throttling before.)

**Comment by brianleroux** [comment]
just added chunking for now / seems better!

**Comment by ryanblock** [comment]
<img src="https://media2.giphy.com/media/PhKhSXofSAm3e/giphy.gif">

**Comment by ryanblock** [comment]
Just following up to commit to memory some unpublished information we received from Amazon as it pertains to throttling and rate limiting Lambda deploys.

Although it's not presently documented, Lambda's management APIs have a hard 15 TPS limit per account.

Some excerpts:

```
- For the Lambda namespace there is a limit of 15 TPS for all control plane APIs (i.e. management APIs, s.a. CreateFunction, GetFunction, UpdateFunctionConfiguration, etc..).
- This rate limit is shared between all these APIs and is per account, meaning that one call to CreateFunction and another call at the same second to UpdateFunctionConfiguration will be counted as 2 calls during that second.
- Unfortunately, this is a hard limit and cannot be increased.
- In order to avoid being throttled by these calls, we recommend distributing calls across the minute as well as implementing retries if calls get throttled. 
- In addition, applying a sleeping time after throttled calls using a backoff algorithm is highly recommended. 
```

Additionally, TPS on Lambda control plane APIs aren't influenced by payload size (i.e. `UpdateFunctionCode` with a 40MB payload isn't treated differently than a 1MB payload)

**Comment by brianleroux** [comment]
Super interesting. We could codify this. Currently we're chunking into 10 deploys per…however long it takes to deploy ten lambdas. Roughly seeing 1mb encodes and uploads within about 1s and, tho I've done no measuring, it stands to reason larger payloads will take longer to upload. So the current deploy script in optimal conditions is within 15/tps. We are probably close to 1/tps. If multiple ppl are performing uploads in parallel they could see problems which, in a larger org, isn't an unlikely scenario. 

Nice thing is we have surgical deploys (deploying one lambda at a time) so if not unlikely at least an uncommon scenario. 

Having these hard limits does invite the opportunity to refactor and see if we can design directly for them however! Currently we run everything in parallel using the Node event loop which schedules things efficiently but we were feeling around in the dark for the limits. Now we know them. We should experiment w spawning chunks of 15 child processes and see how it feels.

**Comment by brianleroux** [comment]
the good news: we have a queue and can now tweak these knobs with a high degree of precision. the meh news: doesn't seem appreciably faster. its probably just more stable now that its tuned to work exactly within the limits amazon described (instead of likely bumping into them and getting throttled but the sdk handled it transparently). closing for now!
